### PR TITLE
[SPARK-29444] Add configuration to support JacksonGenrator to keep fields with null values

### PR DESCRIPTION
**Pull request by jackylee-ch**
### Why are the changes needed?
As mentioned in jira, sometimes we need to be able to support the retention of null columns when writing JSON. 
For example, sparkmagic(used widely in jupyter with livy) will generate sql query results based on DataSet.toJSON and parse JSON to pandas DataFrame to display. If there is a null column, it is easy to have some column missing or even the query result is empty. The loss of the null column in the first row, may cause parsing exceptions or loss of entire column data.

### Does this PR introduce any user-facing change?
Example in spark-shell.
scala> spark.sql("select null as a, 1 as b").toJSON.collect.foreach(println)
{"b":1}                                                                         

scala> spark.sql("set spark.sql.jsonGenerator.struct.ignore.null=false")
res2: org.apache.spark.sql.DataFrame = [key: string, value: string]

scala> spark.sql("select null as a, 1 as b").toJSON.collect.foreach(println)
{"a":null,"b":1}

### How was this patch tested?
Add new test to JacksonGeneratorSuite

### CODE REVIEW

**Comment by stczwd** [comment]
cc @HyukjinKwon @sameeragarwal @xuanyuanking

**Comment by stczwd** [comment]
Previous discussion about this: [SPARK-23773](https://github.com/apache/spark/pull/20884)

**Comment by stczwd** [comment]
cc @cloud-fan

**Comment by dongjoon-hyun** [comment]
Hi, @stczwd . Thank you for making a PR.
Does this PR resolve SPARK-23773 ? Otherwise, you need to file an independent JIRA issue for this.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/json/JacksonGeneratorSuite.scala [hunk 1]
```diff
@@ -39,6 +39,18 @@ class JacksonGeneratorSuite extends SparkFunSuite {
     assert(writer.toString === """{"a":1}""")
   }
 
+  test("initial with StructType and write out an empty row with allowStructIncludeNull=true") {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/json/JacksonGeneratorSuite.scala]: This looks like a test case for bug. We need a JIRA issue ID prefix for the test case name.

Comment (stczwd) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/json/JacksonGeneratorSuite.scala]: ok

**Comment by stczwd** [comment]
> Hi, @stczwd . Thank you for making a PR.
> Does this PR resolve SPARK-23773 ? Otherwise, you need to file an independent JIRA issue for this.

Yes, this PR can resolve this problem

**Comment by stczwd** [comment]
cc @dongjoon-hyun @xuanyuanking @cloud-fan @rdblue 
any other question blocking this merged in？

**Comment by rdblue** [comment]
Sounds fine to me.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala [hunk 2]
```diff
@@ -76,6 +76,9 @@ private[sql] class JSONOptions(
   // Whether to ignore column of all null values or empty array/struct during schema inference
   val dropFieldIfAllNull = parameters.get("dropFieldIfAllNull").map(_.toBoolean).getOrElse(false)
 
+  // Whether to ignore column of all null during json generating
+  val structIngoreNull = parameters.getOrElse("structIngoreNull", "true").toBoolean
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala]: is it specific to struct type column? if not how about naming it `ignroeNullColumns`?

**Comment by cloud-fan** [comment]
The change looks reasonable. Do you know why the json data source ignore null fields at the first place?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/json/JacksonGeneratorSuite.scala [hunk 3]
```diff
@@ -39,6 +39,19 @@ class JacksonGeneratorSuite extends SparkFunSuite {
...
+      "with allowStructIncludeNull=true") {
+    val dataType = StructType(StructField("a", IntegerType) :: Nil)
+    val input = InternalRow(null)
+    val writer = new CharArrayWriter()
+    val allowNullOption =
+      new JSONOptions(Map("structIngoreNull" -> "false"), gmtId)
+    val gen = new JacksonGenerator(dataType, writer, allowNullOption)
+    gen.write(input)
+    gen.flush()
+    assert(writer.toString === """{"a":null}""")
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/json/JacksonGeneratorSuite.scala]: can we also test null inner field? e.g. `{"a": {"b": null}}`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala (hunk 2 above)
Comment (stczwd) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala]: It works on StructType, including struct field and struct inner data

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/json/JacksonGeneratorSuite.scala (hunk 3 above)
Comment (stczwd) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/json/JacksonGeneratorSuite.scala]: Sure, I have added a test for this

**Comment by stczwd** [comment]
> The change looks reasonable. Do you know why the json data source ignore null fields at the first place?

@cloud-fan I don't know about this. Any ideas?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala]: how about top-level columns?

Comment (stczwd) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala]: Yeap, it also works on that

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala]: then shall we pick a better name here?

Comment (stczwd) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala]: okey, ignoreNullFields is much better than structIgnoreNull, I'll change it.

**Comment by stczwd** [comment]
@cloud-fan @dongjoon-hyun 
Any other questions about it?

**Comment by dilipbiswal** [comment]
@stczwd I have a question. After this change when we write to the json file, we will be able to preserve null values. Right ?

```
scala> spark.sql("select null as a, 1 as b").write.format("json").save("/tmp/jsonfile")
```

**Comment by stczwd** [comment]
@dilipbiswal Yes, but case is not right. If you want to preserve null values, you must write like this to disable ignoreNullFields.
```
spark.sql("select null as a, 1 as b").write.format("json").option("ignoreNullFields", "false").save("/tmp/jsonfile")
```

**Comment by dilipbiswal** [comment]
@stczwd Thanks.. sorry .. i missed the newly added option.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 4]
```diff
@@ -1153,6 +1153,12 @@ object SQLConf {
     .booleanConf
     .createWithDefault(true)
 
+  val JSON_GENERATOR_IGNORE_NULL_FIELDS =
+    buildConf("spark.sql.jsonGenerator.nullFields.ignore")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: config name has namespaces, we'd better avoid creating unnecessary namespaces. How about `spark.sql.jsonGenerator.ignoreNullFields`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala [hunk 5]
```diff
@@ -76,6 +76,9 @@ private[sql] class JSONOptions(
   // Whether to ignore column of all null values or empty array/struct during schema inference
   val dropFieldIfAllNull = parameters.get("dropFieldIfAllNull").map(_.toBoolean).getOrElse(false)
 
+  // Whether to ignore null fields during json generating
+  val ignoreNullFields = parameters.getOrElse("ignoreNullFields", "true").toBoolean
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala]: SQLConf should also take effect here. How about
```
parameters.get("ignoreNullFields").map(_.toBoolean).getOrElse {
  SQLConf.get. jsonGeneratorIgnoreNullFields
}
```

Comment (stczwd) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala]: Perfect. Then the code in DataSet.scala is not necessary now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 4 above)
Comment (stczwd) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: okey, I will change it.

**Comment by stczwd** [comment]
@cloud-fan Any more questions?

**Comment by cloud-fan** [comment]
ok to test

**Comment by xuanyuanking** [comment]
The code changes LGTM.
 I suggest adding more context in the PR description, we should clarify this PR is to add config to control the behavior added in SPARK-23773.

**Comment by stczwd** [comment]
> The code changes LGTM.
> I suggest adding more context in the PR description, we should clarify this PR is to add config to control the behavior added in SPARK-23773.
@xuanyuanking Thanks for support. This PR is used to control the default ignorenullfields behavior, not the behavior added in SPARK-23773, which didn't change any thing.

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala [hunk 6]
```diff
@@ -76,6 +77,10 @@ private[sql] class JSONOptions(
   // Whether to ignore column of all null values or empty array/struct during schema inference
   val dropFieldIfAllNull = parameters.get("dropFieldIfAllNull").map(_.toBoolean).getOrElse(false)
 
+  // Whether to ignore null fields during json generating
+  val ignoreNullFields = parameters.getOrElse("ignoreNullFields",
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala]: Hey, you should document this in `DataFrameWrtier`, `DataStreamWrtier`, `readwriter.py`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 7]
```diff
@@ -1153,6 +1153,12 @@ object SQLConf {
     .booleanConf
     .createWithDefault(true)
 
+  val JSON_GENERATOR_IGNORE_NULL_FIELDS =
+    buildConf("spark.sql.jsonGenerator.ignoreNullFields")
+      .doc("If false, JacksonGenerator will generate null for null fields in Struct.")
+      .stringConf
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Why is it a string conf? shouldn't it be a boolean conf?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/json/JacksonGeneratorSuite.scala [hunk 8]
```diff
@@ -39,6 +39,33 @@ class JacksonGeneratorSuite extends SparkFunSuite {
...
+    val fieldType = StructType(StructField("b", IntegerType) :: Nil)
+    val dataType = StructType(StructField("a", fieldType) :: Nil)
+    val input = InternalRow(InternalRow(null))
+    val writer = new CharArrayWriter()
+    val allowNullOption =
+      new JSONOptions(Map("ignoreNullFields" -> "false"), gmtId)
+    val gen = new JacksonGenerator(dataType, writer, allowNullOption)
+    gen.write(input)
+    gen.flush()
+    assert(writer.toString === """{"a":{"b":null}}""")
```
Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/json/JacksonGeneratorSuite.scala]: nit but I would call `close` with a try-catch for a best practice in a followup.

**Comment by HyukjinKwon** [comment]
@stczwd can you make a followup to address the comments above?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala (hunk 6 above)
Comment (stczwd) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala]: Okey, I'll try to follow up this PR, and add this to document

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 9]
```diff
@@ -1153,6 +1153,12 @@ object SQLConf {
     .booleanConf
     .createWithDefault(true)
 
+  val JSON_GENERATOR_IGNORE_NULL_FIELDS =
+    buildConf("spark.sql.jsonGenerator.ignoreNullFields")
+      .doc("If false, JacksonGenerator will generate null for null fields in Struct.")
```
[the remark this answers is not in the document]
Comment (stczwd) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Maybe I can describe this in a better way.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 7 above)
Comment (stczwd) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: It is only used in JasonOptions, where we need string value, and use toBoolean to get boolean conf

**Comment by stczwd** [comment]
Yeah, I'll make it

**Comment by zacikpetr** [comment]
Hi,
Is there some repository I can try this solution?
I would expect it here `https://repository.apache.org/content/repositories/snapshots/`
Thank you for your help.

**Comment by dongjoon-hyun** [comment]
Hi, @zacikpetr .
Yes. It's there. You can try it.
- https://repository.apache.org/content/repositories/snapshots/org/apache/spark/spark-avro_2.12/3.0.0-SNAPSHOT/spark-avro_2.12-3.0.0-20191027.032954-246.jar

**Comment by dongjoon-hyun** [comment]
To make it sure, I triggered the snapshot publishing one minute ago.

**Comment by zacikpetr** [comment]
Thank you @dongjoon-hyun 
Ok, now I am officially confused as I do not understand your build process.

For instance, I am looking for: `sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala`, that should contain `JSON_GENERATOR_IGNORE_NULL_FIELDS`
Class should be part of `spark-catalyst`.

For version https://repository.apache.org/content/repositories/snapshots/org/apache/spark/spark-catalyst_2.11/
only `2.4.5-SNAPSHOT` gets updated
For version https://repository.apache.org/content/repositories/snapshots/org/apache/spark/spark-catalyst_2.12/
only version `3.0.0-SNAPSHOT` gets updated

I tried 2.11 v 3.0.0 and that is unreleased version.
So I try to switch to these updated versions.

I also do not understand how `avro`  is connected to this. 

Best Regards,

**Comment by zacikpetr** [comment]
Can you please also release 3.0.0 for 2.11 as I am stuck with 2.11 version?

**Comment by cloud-fan** [comment]
AFAIK scala 2.11 has been dropped in Spark 3.0, cc @srowen @wangyum

**Comment by srowen** [comment]
This isn't quite the place to discuss, but yes there is no Scala 2.11 support in Spark 3.0.

**Comment by zacikpetr** [comment]
Ok, so keeping context – is there any chance this "issue" can be fixed in Spark 2?

I don't know who is in the lead of this. However, I know that there are currently not many modules supporting Scala 2.12 (spark-streaming-kafka_2.11, spark-csv_2.11, lot of connectors,...), and that makes the transition between Spark 2 and 3 difficult. This situation throws me into the category of  "Spark self workaround developers" to keep Sparking in our project. Is there any plan, recommendation and thread to this topic you can redirect me to?

Thank you all for your time.

**Comment by srowen** [comment]
All of Spark has supported 2.12 since 2.4.0, not sure what you mean.

**Comment by zacikpetr** [comment]
For instance, https://github.com/databricks/spark-csv doesn't seem to support 2.12 version yet. Maybe I just do not understand concept. But I suppose that we cannot combine 2.11 and 2.12 modules.

**Comment by srowen** [comment]
Per the repo, that was long ago pushed into Spark 2.x. CSV parsing is part of Spark and Spark supports 2.12. Third-party packages -- who knows. That memsql one maybe doesn't, but it also hasn't been updated since 2.0.x, so I don't even know if it works on 2.4 + Scala 2.11. All the less on Spark 3.0, likely. As we'll support scala 2.13 in Spark 3.x at some nearish point, we really can't keep supporting 2.11, which is EOL anyway.

**Pull request closed without merging**